Kernel/Storage/ATA/ATAPort.cpp

```diff
@@ -0,0 +1,459 @@
+        // FIXME: Don't assume all drives will have logical sector size of 512 bytes.
+        ATADevice::Address address = { m_port_index, static_cast<u8>(device_index) };
+        m_ata_devices.append(ATADiskDevice::create(m_parent_ata_controller, address, capabilities, 512, max_addressable_block));
+    }
+    return {};
+}
+
+RefPtr<StorageDevice> ATAPort::connected_device(size_t device_index) const
+{
+    MutexLocker locker(m_lock);
+    if (m_ata_devices.size() > device_index)
+        return m_ata_devices[device_index];
+    return {};
+}
+
+ErrorOr<void> ATAPort::start_request(const ATADevice& associated_device, AsyncBlockDeviceRequest& request)
+{
+    MutexLocker locker(m_lock);
+    VERIFY(m_current_request.is_null());
+    VERIFY(pio_capable() || dma_capable());
+
+    dbgln_if(ATA_DEBUG, "ATAPort::start_request");
+
+    m_current_request = request;
+    m_current_request_block_index = 0;
+    m_current_request_flushing_cache = false;
+
+    if (dma_capable()) {
+        TRY(prepare_and_initiate_dma_transaction(associated_device));
+        return {};
+    }
+    TRY(prepare_and_initiate_pio_transaction(associated_device));
+    return {};
+}
+
+void ATAPort::complete_pio_transaction(AsyncDeviceRequest::RequestResult result)
+{
+    VERIFY(m_current_request);
+
+    // Now schedule reading back the buffer as soon as we leave the irq handler.
+    // This is important so that we can safely write the buffer back,
+    // which could cause page faults. Note that this may be called immediately
+    // before Processor::deferred_call_queue returns!
+    g_io_work->queue([this, result]() {
+        dbgln_if(ATA_DEBUG, "ATAPort::complete_pio_transaction result: {}", (int)result);
+        MutexLocker locker(m_lock);
+        VERIFY(m_current_request);
+        auto current_request = m_current_request;
+        m_current_request.clear();
+        current_request->complete(result);
+    });
+}
+
+void ATAPort::complete_dma_transaction(AsyncDeviceRequest::RequestResult result)
+{
+    // NOTE: this may be called from the interrupt handler!
+    VERIFY(m_current_request);
+    VERIFY(m_lock.is_locked());
+
+    // Now schedule reading back the buffer as soon as we leave the irq handler.
+    // This is important so that we can safely write the buffer back,
+    // which could cause page faults. Note that this may be called immediately
+    // before Processor::deferred_call_queue returns!
+    g_io_work->queue([this, result]() {
+        dbgln_if(ATA_DEBUG, "ATAPort::complete_dma_transaction result: {}", (int)result);
+        MutexLocker locker(m_lock);
+        VERIFY(m_current_request);
+        auto current_request = m_current_request;
+        m_current_request.clear();
+
+        if (result == AsyncDeviceRequest::Success) {
+            {
+                auto result = force_busmastering_status_clean();
+                // FIXME: Propagate errors properly
```

Comment:
Can't we just complete the AsyncDeviceRequest with an error here if the `result` is an error?

Reply:
Yes we can actually do that and I'll propagate it to the entire code where possible :)